docs/building-with-stdlib.md

```diff
@@ -0,0 +1,192 @@
+library across ABI boundary - means less code involved and less memcpy. We use
+Standard Library classes used elsewhere in the app. We can optimize the
+event passing by avoiding `KeyValueIterable` transform (and, thus, unnecessary memcpy)
+when we know that the incoming container type matches that one used by SDK.
+
+### Avoiding unnecessary extra memcpy (perf improvements)
+
+No need to transform from 'native' standard library types, e.g. `std::map` via
+`KeyValueIterable` means we can bypass that transformation process, if and when
+we know that the ABI compatibility is not a requirement in certain environment.
+ETA perf improvement is 1.5%-3% better perf since an extra transform, memcpy,
+iteration for-each key-value is avoided. Custom OpenTelemetry SDK implementation
+may operate on standard container classes rather than doing transform from one
+container type to another.
+
+### ABI stability
+
+Obviously this approach does not work in Linux environments, where ABI stability
+guarantee must be provided, e.g. for dynamically loadable OpenTelemetry SDK
+module and plugins for products such as NGINX, Envoy, etc. Example: a product is
+compiled with modern gcc compiler. But OpenTelemetry SDK is compiled with an
+older runtime library. In this case the SDK must be compiled with `nostd`.
+
+Note that for most scenarios with modern Windows compilers, STL library is
+ABI-safe across Visual Studio 2015, 2017 and 2019 with vc14x runtime.
+
+Quote from [official documentation](https://docs.microsoft.com/en-us/cpp/porting/binary-compat-2015-2017?view=msvc-160) :
+
+Visual Studio 2015, 2017, and 2019: the runtime libraries and apps compiled by
+any of these versions of the compiler are binary-compatible. It's reflected in
+the C++ toolset major number, which is 14 for all three versions. The toolset
+version is v140 for Visual Studio 2015, v141 for 2017, and v142 for 2019.
+Say you have third-party libraries built by Visual Studio 2015. You can still
+use them in an application built by Visual Studio 2017 or 2019. There's no need
+to recompile with a matching toolset. The latest version of the Microsoft Visual
+C++ Redistributable package (the Redistributable) works for all of them.
+
+Visual Studio provides 1st class debug experience for the standard library.
+## Build and Test considerations
+
+### Separate flavors of SDK build
+
+Supported build flavors:
+
+- `nostd` - OpenTelemetry backport of classes for C++11. Not using standard lib.
+- `stdlib`   - Standard Library. Full native experience with C++20 compiler.
+  C++17 works but with additional dependencies, e.g. either MS-GSL or Abseil
+  for `std::span` implementation (`gsl::span` or `absl::Span`).
+- `absl`  - TODO: this should allow using Abseil C++ library only (no MS-GSL).
+
+Currently only `nostd` and `stdlib` configurations are implemented.
+`absl` is reserved for future use.
+
+### Build matrix
+
+List of compilers that support building with standard library classes:
```

Comment:
We may also want to note that (for now) this is cmake-only.

Reply:
Abstracting away from concrete build system (CMake or Bazel), it's more like this feature is behind a `feature gate` or `feature flag` controlled by `#define WITH_STL` that is off by default. I can add that to the notes.